src/Serializer.php

```diff
@@ -11,42 +11,84 @@
 
 use Symfony\Component\Serializer\Serializer as BaseSerializer;
 
-final class Serializer extends BaseSerializer
-{
-    private const KEY_TYPE = '#type';
-    private const KEY_SCALAR = '#scalar';
-
-    public function normalize($data, $format = null, array $context = [])
+if (PHP_VERSION_ID >= 80000) {
+    final class Serializer extends BaseSerializer
     {
-        $normalizedData = parent::normalize($data, $format, $context);
+        private const KEY_TYPE = '#type';
+        private const KEY_SCALAR = '#scalar';
+
+        public function normalize($data, $format = null, array $context = []): \ArrayObject|array|string|int|float|bool|null
+        {
+            $normalizedData = parent::normalize($data, $format, $context);
 
-        if (\is_object($data)) {
-            $typeData = [self::KEY_TYPE => \get_class($data)];
-            $valueData = is_scalar($normalizedData) ? [self::KEY_SCALAR => $normalizedData] : $normalizedData;
-            $normalizedData = array_merge($typeData, $valueData);
+            if (\is_object($data)) {
+                $typeData = [self::KEY_TYPE => $data::class];
+                $valueData = is_scalar($normalizedData) ? [self::KEY_SCALAR => $normalizedData] : $normalizedData;
+                $normalizedData = array_merge($typeData, $valueData);
+            }
+
+            return $normalizedData;
         }
 
-        return $normalizedData;
-    }
+        public function denormalize($data, $class, $format = null, array $context = []): mixed
+        {
+            if (\is_array($data) && (isset($data[self::KEY_TYPE]))) {
+                $type = $data[self::KEY_TYPE];
+                unset($data[self::KEY_TYPE]);
 
-    public function denormalize($data, $class, $format = null, array $context = [])
-    {
-        if (\is_array($data) && (isset($data[self::KEY_TYPE]))) {
-            $type = $data[self::KEY_TYPE];
-            unset($data[self::KEY_TYPE]);
+                $data = $data[self::KEY_SCALAR] ?? $data;
+                $data = $this->denormalize($data, $type, $format, $context);
```

Comment:
This line should be removed in the two versions (it is actually one of the fixes introduced in #87), because it triggers two denormalizations, which is wrong in most cases since it will potentially try to denormalize an object.